eth/filters/filter.go

```diff
@@ -147,57 +147,95 @@ func (f *Filter) Logs(ctx context.Context) ([]*types.Log, error) {
 		}
 		return hdr.Number.Int64(), nil
 	}
+
+	var err error
+	// range query need to resolve the special begin/end block number
 	if f.begin, err = resolveSpecial(f.begin); err != nil {
 		return nil, err
 	}
 	if f.end, err = resolveSpecial(f.end); err != nil {
 		return nil, err
 	}
-	// Gather all indexed logs, and finish with non indexed ones
-	var (
-		logs           []*types.Log
-		end            = uint64(f.end)
-		size, sections = f.sys.backend.BloomStatus()
-	)
-	if indexed := sections * size; indexed > uint64(f.begin) {
-		if indexed > end {
-			logs, err = f.indexedLogs(ctx, end)
-		} else {
-			logs, err = f.indexedLogs(ctx, indexed-1)
-		}
-		if err != nil {
-			return logs, err
+
+	logChan, errChan := f.rangeLogsAsync(ctx)
+	var logs []*types.Log
+LOOP:
+	for {
+		select {
+		case log := <-logChan:
+			logs = append(logs, log)
+		case ierr := <-errChan:
+			err = ierr
+			break LOOP
```

Comment:
Instead of the `LOOP` and breaking here, afaict this is the only path to exit. So you could just put "everhthing below line 172` into this switch-clause, and then `return`. ?

Reply:
a good tip, thanks